pkg/async/group.go

```diff
@@ -0,0 +1,62 @@
+package async
+
+import (
+	"context"
+	"fmt"
+)
+
+// A group has several workers, and the group can execute these workers concurrently,
+// wait for the workers to finish within a specified time, and receive the results returned by these workers.
+type Group struct {
+	workers []func(stopCh chan interface{})
+	stopCh  chan interface{}
+	ctx     context.Context
+}
+
+func NewGroup(ctx context.Context) *Group {
+	return &Group{
+		stopCh: make(chan interface{}),
+		ctx:    ctx,
+	}
+}
+
+// Add a worker to group
+func (g *Group) Add(w func(stopCh chan interface{})) {
+	g.workers = append(g.workers, w)
+}
+
+// Execute all workers concurrently, and wait for all workers to end.
+func (g *Group) Wait() []error {
+
+	if g.workers == nil || len(g.workers) == 0 {
+		return nil
+	}
+
+	for _, worker := range g.workers {
+		go worker(g.stopCh)
+	}
+
+	var errs []error
+	res := 0
+	for {
+		select {
+		case <-g.ctx.Done():
+			return []error{
+				fmt.Errorf("time out"),
+			}
+
+		case val := <-g.stopCh:
```

Comment:
stopCh is a blocking channel, if timeout and function return in the first case, worker's sending operation will be blocked and the goroutine will not exit

Reply:
change the stopCh to a buffered channel